pkg/controllers/provisioning/scheduling/nodeset.go

```diff
@@ -76,3 +86,42 @@ func (s *NodeSet) getDaemons(ctx context.Context, constraints *v1alpha5.Constrai
 func (s *NodeSet) Add(node *Node) {
 	s.nodes = append(s.nodes, node)
 }
+
+func (s *NodeSet) Schedule(ctx context.Context, p *v1.Pod) error {
+	// try nodes in ascending order of the number of pods on the node to more evenly distribute nodes
+	// This sorts upon every pod schedule, but it is likely fast enough.  I benchmarked sorting 2000 nodes, 2000 times
+	// which should be an upper bound on our worst case and it took 0.1 seconds.
+	sort.Slice(s.nodes, func(a, b int) bool {
+		return len(s.nodes[a].Pods) < len(s.nodes[b].Pods)
+	})
+	var scheduledNode *Node
+	for _, node := range s.nodes {
+		if err := node.Add(s.topology, p); err == nil {
+			scheduledNode = node
+			break
+		}
+	}
+
+	if scheduledNode == nil {
+		node := NewNode(s.constraints, s.daemonResources, s.instanceTypes)
+		if err := node.Add(s.topology, p); err != nil {
+			return err
+		}
+		s.Add(node)
+		scheduledNode = node
+	}
+
+	if scheduledNode != nil {
+		if err := s.topology.Record(p, scheduledNode.Constraints.Requirements); err != nil {
+			logging.FromContext(ctx).Errorf("Recording topology decision, %s", err)
+			return err
+		}
+	}
+	return nil
+}
+
+type matchingTopology struct {
```

Comment:
Unused now?

Reply:
No, still used: `func (t *Topology) getMatchingTopologies(p *v1.Pod) []matchingTopology`

Reply:
Ah -- just in the wrong file.